core/vm/contracts.go

```
@@ -1203,3 +1216,313 @@ func (c *bls12381MapG2) Run(input []byte) ([]byte, error) {
+	inputGas := uint64(len(input)) * params.ExternalCallByteGasCost
+
+	var callResultGas uint64 = (32 + 32*7) * params.ExternalCallByteGasCost // pay address and topics
+	if maxDataLen%32 != 0 {
+		maxDataLen = maxDataLen + (32 - maxDataLen%32)
+	}
+	callResultGas += maxDataLen * params.ExternalCallByteGasCost
+	gasUsed := callResultGas + inputGas + params.ExternalCallGas
+	/*
+		The gas calculation formula is as follows：
+		gas_overpay =  ExternalCallByteGasCost * (address_len + topics_len + data_len) + ExternalCallGas
+
+		address_len = 32
+		000000000000000000000000751320c36f413a6280ad54487766ae0f780b6f58
+
+		topics_len = 3 * 32 + 32 * 4
+		Note:
+			the first 3 * 32 is the prefix of topics which generated after executing ABIPack(address,topics,data) when the number of topics is 4
+			the last 4 is the maxNumber of topics
+		0000000000000000000000000000000000000000000000000000000000000060
+		00000000000000000000000000000000000000000000000000000000000000c0
+		0000000000000000000000000000000000000000000000000000000000000002
+		dce721dc2d078c030530aeb5511eb76663a705797c2a4a4d41a70dddfb8efca9
+		000000000000000000000000000000000000000000000000000000000000000
+
+		if maxDataLen % 32 != 0
+			data_len= 32 + (32 - maxDataLen % 32) + maxDatalen
+		if maxDataLen % 32 == 0
+			data_len= 32 + maxDatalen
+		Note:
+			the first 32 is the length of data which generated after ABIPack(address,topics,data)
+		00000000000000000000000000000000000000000000000000000000000000c0
+		0000000000000000000000000000000000000000000000000000000000000002
+		0000000000000000000000000000000000000000000000000000000000000001
+		0000000000000000000000000000000000000000000000000000000000000060
+		0000000000000000000000000000000000000000000000000000000000000028
+		bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb
+		bbbbbbbbbbbbbbbb000000000000000000000000000000000000000000000000
+	*/
+
+	return gasUsed
+}
+
+func (c *crossChainCall) Run(input []byte) ([]byte, error) {
+	return nil, nil
```

Prefer to return an explicit error since it's not supposed to be called